rails_event_store/lib/rails_event_store/active_job_dispatcher.rb

```diff
@@ -0,0 +1,64 @@
+require 'active_job'
+
+module RailsEventStore
+  module AsyncProxyStrategy
+    class AfterCommit
+      def call(klass, event)
+        if ActiveRecord::Base.connection.transaction_open?
+          ActiveRecord::Base.
+            connection.
+            current_transaction.
+            add_record(AsyncRecord.new(klass, event))
+        else
+          klass.perform_later(YAML.dump(event))
+        end
+      end
+
+      private
+      class AsyncRecord
+        def initialize(klass, event)
+          @klass = klass
+          @event = event
+        end
+
+        def committed!
+          @klass.perform_later(YAML.dump(@event))
+        end
+      end
+    end
+
+    class Inline
+      def call(klass, event)
+        klass.perform_later(YAML.dump(event))
+      end
+    end
+  end
+
+  class ActiveJobDispatcher
+    def initialize(proxy_strategy: AsyncProxyStrategy::Inline.new)
+      @async_proxy_strategy = proxy_strategy
+    end
+
+    def call(subscriber, event)
+      subscriber.call(event)
+    end
+
+    def proxy_for(klass)
+      async_handler?(klass) ? async_proxy(klass) : sync_proxy(klass)
+    end
```

Comment:
Wouldn't it be a simpler (in terms of code and understanding) to decide in `Dispatcher#call` whether `subscriber` needs to wrapped in proxy?

Reply:
Right now it happens in event broker. Broker checks if proxy is needed and delegates to dispatcher to wrap. Dispatcher returnes wrapped subscriber and then receives it in `#call`.

I could see that current implementation results in fewer calls but it took me a while to grasp it. Is performance on this code path significant? (benchmark neeeded)